TimeoutPersister query operators fixed

Scooletz opened the pull request:
This PR changes the operators for querying timeout bringing it back to `>=` and `<=`.

This PR fixes the operators to its original shape before https://github.com/Particular/NServiceBus.Persistence.AzureStorage/commit/240e24e5e849981cbc022ff1279ba0015998ee91

Somehow connected to #118

PoA:
- [x] fix operators
- [x] ~possibly remove `TimeoutManagerDataEntity` as it looks like uneeded. It might be just a field in the timeout manager restored at the start of the endpoint. If needed, provide reasoning.~
- [x] move storing the cursor (`LastSuccesfulRead`) to the next spin of `GetChunk` to not move cursor beyond timeouts that weren't acknowledged yet.

yvesgoeleven commented on src/NServiceBus.Persistence.AzureStorage/Timeout/TimeoutLogic/TimeoutPersister.cs:
```diff
@@ -164,9 +164,9 @@ public TimeoutPersister(string timeoutConnectionString, string timeoutDataTableN
                         TableQuery.CombineFilters(
                             TableQuery.CombineFilters(
-                                TableQuery.GenerateFilterCondition("PartitionKey", QueryComparisons.NotEqual, now.ToString(partitionKeyScope)),
+                                TableQuery.GenerateFilterCondition("PartitionKey", QueryComparisons.LessThan, now.ToString(partitionKeyScope)),
```
it should be an inclusive range, so less than or equal (or less than the next pkey)

yvesgoeleven commented on src/NServiceBus.Persistence.AzureStorage/Timeout/TimeoutLogic/TimeoutPersister.cs:
```diff
@@ -164,9 +164,9 @@ public TimeoutPersister(string timeoutConnectionString, string timeoutDataTableN
+                                TableQuery.GenerateFilterCondition("PartitionKey", QueryComparisons.LessThan, now.ToString(partitionKeyScope)),
                                 TableOperators.And,
-                                TableQuery.GenerateFilterCondition("PartitionKey", QueryComparisons.NotEqual, lastSuccessfulRead.Value.ToString(partitionKeyScope))),
+                                TableQuery.GenerateFilterCondition("PartitionKey", QueryComparisons.GreaterThan, lastSuccessfulRead.Value.ToString(partitionKeyScope))),
```
it should be an inclusive range, so greater than or equal (or greater than the previous pkey)

Scooletz commented:
Storing `TimeoutManagerDataEntity` is required. Scanning whole table even consisting of many partitions, with the segmented query, could result in a very long operation (imagine 1 millions timeouts stored and the query returning one by one). Storing the cursor is needed then. 
The fact that we store many cursors (one per instance) may result in duplicates, but it's not that likely because of the latencies and different startup times.

I've updated the PoA with a point about moving cursor and storing the change.

Scooletz commented:
Ping @Particular/azure-maintainers 
I moved the update of the last successful read to the next spin. This addresses all the issues related to this PR.

SeanFeldman commented on src/NServiceBus.Persistence.AzureStorage/Timeout/TimeoutLogic/TimeoutPersister.cs:
```diff
@@ -164,9 +166,9 @@ public TimeoutPersister(string timeoutConnectionString, string timeoutDataTableN
                         TableQuery.CombineFilters(
                             TableQuery.CombineFilters(
-                                TableQuery.GenerateFilterCondition("PartitionKey", QueryComparisons.NotEqual, now.ToString(partitionKeyScope)),
+                                TableQuery.GenerateFilterCondition("PartitionKey", QueryComparisons.LessThanOrEqual, now.ToString(partitionKeyScope)),
```
For the readability, since this code is changing, would it improve readability if to swap the two conditions?
Instead of reading it as  
`PK <= now AND PK >= lastSuccessfulRead`
to have
`PK >= lastSuccessfulRead AND PK <= now`

SeanFeldman commented on src/NServiceBus.Persistence.AzureStorage/Timeout/TimeoutLogic/TimeoutPersister.cs:
```diff
@@ -153,6 +153,8 @@ public TimeoutPersister(string timeoutConnectionString, string timeoutDataTableN
             var timeoutDataTable = client.GetTableReference(timeoutDataTableName);
             var timeoutManagerDataTable = client.GetTableReference(timeoutManagerDataTableName);
 
+            await TryUpdateSuccessfulRead(timeoutManagerDataTable);
```
ConfigureAwait?

boblangley commented:
After pairing with @SeanFeldman today and further investigation I am convinced the change to the query operator is only a performance improvement/minor bug

The two `NotEquals` operators were incorrectly filtering the date formatted primary keys, and the change addresses that. However the GUID primary keys are not necessarily filtered, so this would essentially severely hamper any attempted culling based on Primary Keys since it doesn't address those primary keys as well. 

There is a remote chance of a bug that would cause a timeout to be skipped, but that would only occur if both the SagaId and the MessageId GUIDs start with the same characters as the Timeouts trigger time partition formatted string. Otherwise one or both of those records would still be returned and the rest of the operation would be successful. 

To make the culling slightly more aggressive to I added PR #121 which targets this branch. It may be more than we need, but it attempts to help with the GUID primary key culling.

If we want to make this query ultimately do what we intended we should (re)consider splitting the data between 3 tables, one for each primary key entity (time, messageid, sagaid). As long as those other keys are in there we cannot prevent extra results, but that may be a performance optimization too far, would require migration (yuck), and likely has other tradeoffs I have not considered.

It seems that the change to the `lastSuccessfullRead` behavior is meant to addresses #118? Would it be make sense to split that efforts into a seperate PR?

SeanFeldman commented:
Pulling this in. @Particular/azure-maintainers please review if #121 should be also added as it could reduce number of records fetched from timeouts table and uses a single query regardless if there was a successful previous read or not.
